### ISSUE TITLE
tabstop will be set to 8 when I'm editing a .py file

### ISSUE THREAD

**Issue by jiajunhuang** [neovim/neovim#3786]
I've set tabstop to 4 in `~/.config/nvim/init.vim`,  and it works fine in most situations.  but when I'm editing a 
python file, it turns to 8. that's strange.

**Comment by mhinz** [comment]
See `:e $VIMRUNTIME/ftplugin/python.vim`.

It sets:
```viml
" As suggested by PEP8.
setlocal expandtab shiftwidth=4 softtabstop=4 tabstop=8
```

If you want to avoid that, put this in your init.vim:

```viml
autocmd FileType python setlocal tabstop=4
```

**Comment by ZyX-I** [comment]
@mhinz I would not say that overriding user settings with random values because of some code standard that user does not have to follow is desirable. This problem was already discussed in vim mailing lists and Bram did not want to change anything, we may have another idea on this matter.

**Comment by jiajunhuang** [comment]
agree with @ZyX-I , I think the editor shoudl put user's configuration in the first place.

**Comment by jiajunhuang** [comment]
by the way, I've not found these settings in vim 7.4 compiled myself. below is the compile message:

```
Compilation: gcc -c -I. -Iproto -DHAVE_CONFIG_H     -g -O2 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1      
Linking: gcc   -L/usr/local/lib -Wl,--as-needed -o vim        -lm -ltinfo -lnsl  -lselinux     -L/usr/local/lib/python2.7/config -lpython2.7 -lpthread -ldl -lutil -lm -Xlinker -export-dynamic
```

**Comment by justinmk** [comment]
If vim_dev won't change this we can perhaps make an exception and change it here.

**Comment by justinmk** [comment]
Well we are behind on runtime updates...

**Comment by chrisbra** [comment]
> @mhinz I would not say that overriding user settings with random
> values because of some code standard that user does not have to follow
> is desirable. This problem was already discussed in vim mailing lists
> and Bram did not want to change anything, we may have another idea on
> this matter.

This is not entirely correct. There already exist several documented 
methods on how to keep specific filetype settings (:h ftplugin-overrule)
or simply turn off filetype plugins  or use the new OptionSet 
autocommand. I don't see what problem you are trying to address here.

**Comment by ZyX-I** [comment]
@chrisbra You are missing the point. User in the vimrc has told vim that he prefers specific tabstop value, and then Python filetype plugin jumps in and says that Python files should use another value. The problem is not that *there is no way to set needed value*. The problem is that *existing configuration is ignored* (and needs some hacks to overrule this).

**Comment by chrisbra** [comment]
Hi Nikolai!

On So, 06 Dez 2015, Nikolai Aleksandrovich Pavlov wrote:

> @chrisbra You are missing the point.

I don't think so. 

> User in the vimrc has told vim
> that he prefers specific tabstop value, and then Python filetype
> plugin jumps in and says that Python files should use another value.
> The problem is not that *there is no way to set needed value*. The
> problem is that *existing configuration is ignored* (and needs some
> hacks to overrule this).

As I said already, if the user wants to keep his value, than he should 
not enable filetype plugins. This is working as expecting, considering 
that we are talking about buffer-local settings.

Plus I already mentioned several ways to work around that.

**Comment by ZyX-I** [comment]
@chrisbra Filetype plugins are for things like completion, &comments\*, overriding “go to definition” mapping, etc. This stuff *is* filetype-specific. There is no reason to have settings that are in first place *project*-specific and in second a matter of user preference in a filetype plugin.

**Comment by chrisbra** [comment]
> @chrisbra Filetype plugins are for things like completion,
> &comments\*, overriding “go to definition” mapping, etc.

I wouldn't go so far to say some settings shouldn't be set in a filetype 
and it is not documented that way and if you don't like certain options 
in filetype plugins then there are documented ways to achieve what one 
wants (as have been shown in this issue).

> This stuff *is* filetype-specific. There is no reason to have settings
> that are in first place *project*-specific and in second a matter of
> user preference in a filetype plugin.

Well, we are obviously not agreeing here. Your project, your rules. I 
have expressed my opinion and won't argue further and will shut up then 
again.

**Comment by jiajunhuang** [comment]
@ZyX-I shouldn't a editor behave like this: we have predefined rules for filetype. but if user has override some rules, we put the user-defined rules in the first place, else we use the pre-defined rules.

**Comment by ZyX-I** [comment]
@jiajunhuang This could be optimal, but currently there is no way to determine this except for the variant with parsing `:redir … | set optname? | redir END`, and even this does not work in some cases. Needs either a new command (e.g. `setlocaldefault`) or function like `optionsid(optname) -> SID|0|"modeline"|"user"` which returns script ID where given option was defined (0 for options defined by Vim, `"modeline"` for options set from modeline, `"user"` for options set by user in the command-line, by mappigs or by anything else which sets an option, but does not assign SID to it).

**Comment by zhou13** [comment]
In fact vim can have user defined rules for filetype.  Just put your configuration file under $VIMRUNTIME/after/ftplugin.  However, I would say that such overriding problem happens everywhere in VIM. Even the official ftplugins mixes the usage of set/setlocal I think. It will be better if neovim provides a standard that how a ftplugin should provide the default settings for such configuration.

**Comment by black-puppydog** [comment]
So to override this setting (a also found this issue due to python indentation, I indent my code with two spaces) I would have to create another file in $VIMRUNTIME?
That seems pretty bulky to me to be honest.

I wouldn't consider editing a full copy of the python file as a real solution because of the obvious update problem.
So if I want to setup a new pc with neovim I have to either copy/symlink a file into the new install folder (and make sure to not accidentally delete it when I recompile for some reason) or I need to modify the $VIMRUNTIME before starting neovim? To get it to respect a setting I already have packed neatly into my config?

IMHO this not only is highly unintuitive, it is also kind of tricky to debug (for a beginner) given that "enabling filetype plugins" is not something I explicitly chose to do. Heck, in neovim I didn't even do it *implicitly* by enabling syntax highlighting or some such.

So best I can tell is that this is *by default* ignoring an admittedly broad but explicit user setting...

**Comment by black-puppydog** [comment]
Okay, following to [this SO answer](http://stackoverflow.com/a/21074056/393885) you can create the extra file in a .config/nvim subfolder and even though `:echo $VIMRUNTIME` does not list that folder for me it seems to work.
It is still bulky but it is definitely better than it sounded from reading `:h ftplugin-overrule`

**Comment by KillTheMule** [comment]
Aren't aucmds run after the ftplugins are run? So you'll simply not only `set stuff` in your vimrc (because that only sets stuff at the beginning of your editing session), but somehing like `aucmd BufNew,BufNewFile &ft=python set stuff`? Did work for me with a short try.

**Comment by jamessan** [comment]
@black-puppydog `$VIMRUNTIME` is just the system runtime directory.  If you want to see all the runtime directories that (neo)vim will look at, use `:set runtimepath?`.  You'll see there that it does have the user's directory included.

The canonical way to override a setting in the system runtime files is to put a corresponding file in `~/.config/nvim/after` with the overrides you want.  This is the entire reason the after directory exists.  In the case presented by the reporter, that would be putting `setlocal tabstop=4` in `~/.config/nvim/after/ftplugin/python.vim`.

For what it's worth, I'm surprised there's so much push-back for providing sane defaults (i.e. the community preferred style) for a language, since that seems to align with Neovim's goal of providing sane defaults to users.  I agree with the original closing of this issue and don't see why Neovim should deviate from upstream on this.

**Comment by black-puppydog** [comment]
@jamessan I am absolutely not opposed to sane defaults, really, but the docs mentioned VIMRUNTIME, which I echoed and it only includes my installation directory of neovim, and I didn't want to put config in there or override it by making a non-updating copy.
The `/after` directory did not stand out to me as something canonical at all (although now, reading `:h rtp` it becomes clearer) but rather like some directory that the author happened to call "after" and it seemed to me like I would have to make sure to put it at the *end* of VIMRUNTIME, and that struck me as weird.

**Comment by fmoralesc** [comment]
@black-puppydog By default, the `after` dir is always the last entry in `&runtimepath`, and any decent plugin/rtp manager will try to keep it that way.

**Comment by rosshadden** [comment]
I ended up putting this in my vimrc, within an `augroup python` block:

```vim
	let ts = &tabstop
	autocmd FileType python exec 'setlocal tabstop=' . ts
```

**Comment by jamessan** [comment]
There's a [runtime update] in Vim which provides a variable to control this behavior of the ftplugin.  Once that's merged, people can let it if they don't want PEP8 conformant configuration by default.

[runtime update]: https://github.com/vim/vim/commit/abd468ed0fbcba391e7833feeaa7de3ced841455

**Comment by skewerr** [comment]
@jamessan mentioned it already, and it's been in upstream Vim for a while now, so why hasn't this been added to Neovim?

Vim's `ftplugin/python.vim`:

	if !exists("g:python_recommended_style") || g:python_recommended_style != 0
	    " As suggested by PEP8.
	    setlocal expandtab shiftwidth=4 softtabstop=4 tabstop=8
	endif

Neovim's `ftplugin/python.vim`:

	" As suggested by PEP8.
	setlocal expandtab shiftwidth=4 softtabstop=4 tabstop=8

I have `let g:python_recommended_style = 0` in my `.vimrc` and `init.vim`, so I found it strange when I started using Neovim and got expanded-to-4-spaces tabs on my python scripts.

**Comment by justinmk** [comment]
Fixed by 2648c3579a4d (Nvim 0.2.0)
